chapters/Beacons.md

```diff
@@ -0,0 +1,587 @@
+
+### Processing Signals
+
+Given:
+
+* `did` - required, the **did:btc1** identifier whose signals are to be processed
+* `sidecarDocuments` - required, array of documents required for resolution not stored on a ::CAS::,including:
+  * Initial DID document, if `did` was constructed with `idType` of "external" (`did` has the form "did:btc1:x1...")
+  * Map documents for ::BTC1 Beacons:: with services of `beaconType` "MapBeacon".
+  * DID update payloads for each ::BTC1 Beacon Announcement::
+* `smtProofs` - required for services of `beaconType` "SMTBeacon", array of SMT proofs of inclusion or non-inclusion
+* `cas` - optional, one of:
+    * "ipfs"
+* `targetVersionId` - optional, DID document version ID required
+* `targetVersionTime` - optional, DID document version time required
+
+Process the ::Beacon Signals:: to reconstruct the DID document:
+
+1. If `cas` is defined and is not a valid value per above, raise InvalidParameter error.
+1. Set `sidecarDocumentsMap` to empty map.
+1. For each `sidecarDocument` in `sidecarDocuments`:
+   1. Set `hashBytes` to the result of passing `sidecarDocument` to the [JSON Canonicalization and Hash] algorithm.
+   1. Set `id` to the hexadecimal string representation of `hashBytes`.
+   1. Add `id` and `sidecarDocument` to `sidecarDocumentsMap`.
+1. Set `smtProofsMap` to empty map.
+1. For each `smtProof` in `smtProofs`:
+    1. Add `smtProof.id` and `smtProof` to `smtProofsMap`.
+1. If `did` was constructed with `idType` "key", set `didDocument` to the deterministically generated initial document.
+1. If `did` was constructed with `idType` "external":
+   1. Extract `genesisBytes` from `did`.
+   1. Set `id` to the hexadecimal string representation of `genesisBytes`.
+   1. Get `didDocument` from `sidecarDocumentsMap` by its `id` if available, or from ::CAS:: by its `id` if not and `cas` is defined.
+   1. Update placeholder values in `didDocument` with `did` as required.
+1. Until terminated:
+   1. Set `didUpdatePayload` to null.
+   1. For each `service` in `didDocument.service` where `service.type` is "BTC1Beacon":
+      1. Get the next transaction from the Bitcoin address at `service.serviceEndpoint`.
+      1. If `targetVersionTime` is defined and is less than the transaction time, skip to next `service`.
+      1. If the last output transaction is not of the form `[OP_RETURN, OP_PUSHBYTES32, <hashBytes>]`, skip to next `service`.
+      1. Extract `hashBytes` from the last output transaction.
+      1. If `service.beaconType` is not "SingletonBeacon", "MapBeacon", or "SMTBeacon", raise InvalidParameter error.
+      1. If `service.beaconType` is "SingletonBeacon", set `tempDidUpdatePayload` to the result of [Process Singleton Beacon Signal].
+      1. If `service.beaconType` is "MapBeacon", set `tempDidUpdatePayload` to the result of [Process Map Beacon Signal].
+      1. If `service.beaconType` is "SMTBeacon", set `tempDidUpdatePayload` to the result of [Process SMT Beacon Signal].
```

Comment:
This is fine for now. But I will note this is still under discussion and may need fixing later. Shouldnt stop this getting in